// _eschecker@0.0.5@eschecker/lib/semantic-model/types/string-type.js
import Immutable from "immutable";
import assert from "assert";

import {RecordType} from "./record-type";
import {NumberType} from "./number-type";
import {FunctionType} from "./function-type";
import {MaybeType} from "./maybe-type";
import {BooleanType} from "./boolean-type";

let builtIns;

/**
 * string type
 */
export class StringType extends RecordType {

	static create() {
		return instance;
	}

	get prettyName() {
		return "string";
	}

	fresh() {
		return this;
	}

	hasProperty(symbol) {
		return builtIns.has(symbol.name);
	}

	getType(symbol) {
		return builtIns.get(symbol.name);
	}

	setType() {
		assert.fail("Cannot modify properties of the built in type string");
	}
}

const instance = new StringType();

builtIns = new Immutable.Map([
	["length", NumberType.create()],
	["charAt", new FunctionType(instance, [NumberType.create()], instance)],
	["charCodeAt", new FunctionType(instance, [NumberType.create()], NumberType.create())],
	["codePointAt", new FunctionType(instance, [NumberType.create()], NumberType.create())],
	// TODO concat requires varargs
	// ["concat", new FunctionType(stringT, [numberT], numberT)]
	["endsWith", new FunctionType(instance, [instance, MaybeType.of(NumberType.create())], BooleanType.create())],
	["includes", new FunctionType(instance, [instance, MaybeType.of(NumberType.create())], BooleanType.create())],
	["indexOf", new FunctionType(instance, [instance, MaybeType.of(NumberType.create())], NumberType.create())],
	["lastIndexOf", new FunctionType(instance, [instance, MaybeType.of(NumberType.create())], NumberType.create())],
	// TODO match requires regex
	["normalize", new FunctionType(instance, [MaybeType.of(instance)], instance)], // TODO from is an enum
	["repeat", new FunctionType(instance, [NumberType.create()], instance)],
	["replace", new FunctionType(instance, [instance, instance], NumberType.create())], // TODO old can be regex or string, new can be string or function, requires union
	// TODO search requires regex
	["slice", new FunctionType(instance, [NumberType.create(), MaybeType.of(NumberType.create())], instance)],
	// TODO fails because of cycels in import ["split", new FunctionType(instance, [MaybeType.of(instance), MaybeType.of(NumberType.create())], ArrayType.of(instance))]
	["startsWith", new FunctionType(instance, [instance, MaybeType.of(NumberType.create())], BooleanType.create())],
	["substr", new FunctionType(instance, [instance, MaybeType.of(NumberType.create())], instance)],
	["substring", new FunctionType(instance, [NumberType.create(), MaybeType.of(NumberType.create())], instance)],
	["toLocaleLowerCase", new FunctionType(instance, [], instance)],
	["toLocaleUpperCase", new FunctionType(instance, [], instance)],
	["toLowerCase", new FunctionType(instance, [], instance)],
	["toString", new FunctionType(instance, [], instance)],
	["toUpperCase", new FunctionType(instance, [], instance)],
	["trim", new FunctionType(instance, [], instance)],
	["valueOf", new FunctionType(instance, [], instance)]
	// TODO string.raw
]);

export default StringType;
