// _eschecker@0.0.5@eschecker/es5/lib/type-inference/type-unificator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.TypeUnificator = exports.NotUnifiableError = exports.UnificationError = undefined;

var _immutable = require("immutable");

var _immutable2 = _interopRequireDefault(_immutable);

var _util = require("../util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Interface for a base type unification rule.
 * A rule implements two methods. The first method indicates if this rule can handle the given type tuple.
 * If this is the case, then the unification method is responsible for unifiyng the given types.
 *
 * @typedef {Object} BaseTypeUnificationRule
 * @interface
 *
 * @property {function (t1: Type, t2: Type): boolean} canUnify returns true if this rule can handle the given type tuple, false otherwise
 * @property {function (t1: Type, t2: Type, unificator: TypeUnificator): Type} unify unifies the given to types into the most common type and returning this most common type
 */

/**
 * Error thrown if the unification of two types has failed
 */
class UnificationError extends Error {
	constructor(t1, t2, reason) {
		super(`Unification for type '${ t1 }' and '${ t2 }' failed because ${ reason }.`);
		/**
   * One of the types that has been passed to the unification
   * @type {Type}
   */
		this.t1 = t1;

		/**
   * The other type that has been passed to the unification
   * @type {Type}
   */
		this.t2 = t2;
	}
}

exports.UnificationError = UnificationError;
class NotUnifiableError extends UnificationError {
	constructor(t1, t2) {
		super(t1, t2, "there exists no rule that can be used to unify the given types");
	}
}

exports.NotUnifiableError = NotUnifiableError; /**
                                                * Generic implementation of the Hindley Milner algorithm.
                                                * The class accepts a set of rules that should be applied to unify two base types.
                                                * For each type tuple, exactly one rule can handle the unification.
                                                */

class TypeUnificator {

	/**
  * Creates a new instance that uses the passed in unification rules or the default unification rules if no
  * unification rules have been passed in
  * @param {BaseTypeUnificationRule[]} [baseTypeUnificationRules] the unification rules to use
     */
	constructor() {
		let baseTypeUnificationRules = arguments.length <= 0 || arguments[0] === undefined ? (0, _util.globRequireInstances)("./unification-rules/*.js", module) : arguments[0];

		this.baseTypeUnificationRules = _immutable2.default.Set.of(...baseTypeUnificationRules);
	}

	/**
  * Tries to unify the two types and returns the unified type.
  * @param {Type} t1 the first of the types to unify
  * @param {Type} t2 the second of the types to unify
  * @returns {Type} the unified type
  * @throws UnificationError if the unification of the two types is not possible
     */
	unify(t1, t2) {
		// in case the two types are equal, then nothing is to be done
		if (t1.equals(t2)) {
			return t1;
		}

		// both types are base types, base types need to match so that they can be unified
		if (t1.isBaseType && t2.isBaseType) {
			return this._unifyBaseTypes(t1, t2);
		}

		if (t1.isTypeVariable) {
			if (t1.occursIn(t2)) {
				throw new UnificationError(t1, t2, "The type variable of t1 is contained inside of the type t2 and therefore cannot be replaced by t2");
			}

			return t2;
		}

		// t2 is a type variable, t1 is not, switch the type arguments so that t1 is a type variable and t2 is the base type
		// in this case the rule directly above applies
		return this.unify(t2, t1);
	}

	/**
  * Tries to unify the given base types by applying one of the base type unification rules.
  * @param {Type} t1 the first base type
  * @param {Type} t2 the second base type
  * @returns {Type} the unified type
     * @private
     */
	_unifyBaseTypes(t1, t2) {
		return this._getUnificationRuleOrThrow(t1, t2).unify(t1, t2, this);
	}

	_getUnificationRuleOrThrow(t1, t2) {
		let possibleRules = this.baseTypeUnificationRules.filter(rule => rule.canUnify(t1, t2)).toArray();

		switch (possibleRules.length) {
			case 0:
				throw new NotUnifiableError(t1, t2);
			case 1:
				return possibleRules[0];
			default:
				var ruleNames = possibleRules.map(rule => rule.constructor.name).join();
				throw new UnificationError(t1, t2, `unification rule to use is ambiguous(${ ruleNames })`);
		}
	}
}

exports.TypeUnificator = TypeUnificator;
exports.default = TypeUnificator;
